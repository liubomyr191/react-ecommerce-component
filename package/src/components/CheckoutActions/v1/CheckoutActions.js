import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { withComponents } from "@reactioncommerce/components-context";
import { applyTheme, CustomPropTypes } from "../../../utils";

const Action = styled.div`
  border-bottom: solid 1px ${applyTheme("color_black10")};
  padding: 1rem 0;

  &:first-of-type {
    border-top: solid 1px ${applyTheme("color_black10")};
  }
`;

const FormActions = styled.div`
  align-items: flex-end;
  display: flex;
  justify-content: space-between;
  padding: 1rem 0;
`;

class CheckoutActions extends Component {
  static propTypes = {
    actions: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string.isRequired,
      component: CustomPropTypes.component.isRequired,
      props: PropTypes.shape({
        cartData: PropTypes.object,
        cartMutation: PropTypes.func
      })
    })),
    /**
     * If you've set up a components context using @reactioncommerce/components-context
     * (recommended), then this prop will come from there automatically. If you have not
     * set up a components context or you want to override one of the components in a
     * single spot, you can pass in the components prop directly.
     */
    components: PropTypes.shape({
      /**
       * Pass either the Reaction Button component or your own component that
       * accepts compatible props.
       */
      Button: CustomPropTypes.component.isRequired,
      /**
       * Pass either the Reaction CheckoutAction component or your own component that
       * accepts compatible props.
       */
      CheckoutAction: CustomPropTypes.component.isRequired,
      /**
       * Pass either the Reaction CheckoutActionComplete component or your own component that
       * accepts compatible props.
       */
      CheckoutActionComplete: CustomPropTypes.component.isRequired,
      /**
       * Pass either the Reaction CheckboxActionIncomplete component or your own component that
       * accepts compatible props.
       */
      CheckoutActionIncomplete: CustomPropTypes.component.isRequired
    }).isRequired
  };

  static defaultProps = {};

  state = {
    currentActions: this.props.actions.map(({ label, props }, i) => ({
      label,
      // eslint-disable-next-line
      status: props.cartData.data ? "complete" : i === 0 ? "active" : "incomplete",
      readyForSave: false,
      isSaving: false,
      capturedData: null
    }))
  };

  getCurrentActionIndex(label) {
    const { currentActions } = this.state;
    return currentActions.findIndex((action) => action.label === label);
  }

  getCurrentActionByLabel(label) {
    const { currentActions } = this.state;
    return currentActions[this.getCurrentActionIndex(label)];
  }

  toggleActionStatus = (label, status) => {
    const { currentActions } = this.state;
    const actionIndex = currentActions.findIndex((action) => action.label === label);
    currentActions[actionIndex].status = status;
    this.setState({ currentActions });
  };

  actionReadyForSave = (label, ready) => {
    const { currentActions } = this.state;
    currentActions[this.getCurrentActionIndex(label)].readyForSave = ready;
    this.setState({
      currentActions
    });
  };

  captureActionData = async ({ label, props: { cartMutation } }) => {
    const actionData = await this[label].getFullfillmentData();
    const { currentActions } = this.state;
    currentActions[this.getCurrentActionIndex(label)].isSaving = true;
    this.setState({
      currentActions
    });

    cartMutation(actionData).then(() => {
      currentActions[this.getCurrentActionIndex(label)].isSaving = false;
      currentActions[this.getCurrentActionIndex(label)].status = "complete";
      this.setState({
        currentActions
      });
    });
  };

  renderCompleteAction = ({ label, component, props: { cartData } }) => {
    const { components: { CheckoutActionComplete } } = this.props;
    return cartData.data ? (
      <CheckoutActionComplete
        key={label}
        content={component.renderComplete(cartData.data)}
        onClickChangeButton={() => {
          this.toggleActionStatus(label, "active");
        }}
      />
    ) : (
      <span />
    );
  };

  renderActiveAction = ({ component: Comp, ...action }) => {
    const { components: { Button } } = this.props;
    const { readyForSave, isSaving } = this.getCurrentActionByLabel(action.label);

    return (
      <Fragment>
        <Comp
          value={action.props.cartData.data}
          onReadyForSaveChange={(ready) => {
            this.actionReadyForSave(action.label, ready);
          }}
          isSaving={isSaving}
          ref={(el) => {
            this[action.label] = el;
          }}
        />
        <FormActions>
          {action.props.cartData.data ? (
            <Button
              actionType="secondary"
              isTextOnly
              isShortHeight
              onClick={() => this.toggleActionStatus(action.label, "complete")}
            >
              Cancel
            </Button>
          ) : (
            ""
          )}
          <Button
            actionType="important"
            isShortHeight
            onClick={() => this.captureActionData(action)}
            isDisabled={!readyForSave}
          >
            Save and continue
          </Button>
        </FormActions>
      </Fragment>
    );
  };

  renderAction = (action) => {
    const { components: { CheckoutAction, CheckoutActionIncomplete } } = this.props;
    const { status } = this.getCurrentActionByLabel(action.label);

    return (
      <Action key={action.label}>
        <CheckoutAction
          status={status}
          label={action.label}
          stepNumber={this.getCurrentActionIndex(action.label) + 1}
          activeStepElement={this.renderActiveAction(action)}
          completeStepElement={this.renderCompleteAction(action)}
          incompleteStepElement={<CheckoutActionIncomplete />}
        />
      </Action>
    );
  };

  render() {
    const { actions } = this.props;
    return actions.map(this.renderAction);
  }
}

export default withComponents(CheckoutActions);
