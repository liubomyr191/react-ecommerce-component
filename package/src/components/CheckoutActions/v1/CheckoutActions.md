### Overview
The `CheckoutActions` component is responsible for:
  * Displaying each `CheckoutAction` and managing it's status. (complete, incomplete, active)
  * Providing each  `CheckoutAction` with any data it may need from the `cart` object.
  * Capturing/Editing of `CheckoutAction` data.
  * Rendering captured `CheckoutAction` data in a `CheckoutActionComplete` component.

#### Usage
`CheckoutActions` take an array of `actions`, each `action` needs a `label` and a checkout action component that will be responsible to capturing a piece of checkout data as well as a `props` object to include the piece of `cart.checkout` data the action needs to display and a `onSubmit` funciton that needs to be called during action submission.

**Example of Actions array**
```js static
const actions = [
  {
    label: "Shipping Information",
    component: ShippingAddressCheckoutAction,
    onSubmit: setShippingAddress
    props: cart.checkout.fulfillmentGroup
  },
  {
    label: "Shipping Options",
    component: ShippingOptionCheckoutAction,
    onSubmit: setShippingOption
    props: cart.checkout.fulfillmentGroup.avalibleFulfilmentGroups
  },
  { 
    label: "Payment Information", 
    component: PaymentCheckoutAction,
    onSubmit: setPayment,
    props: cart.checkout.payments[0]
  }
];

```

**Note:** These examples only use the `ShippingAddressCheckoutAction` as the actions component. This will be updated with more actions as they get created.

```jsx
initialState = { cart: null }
const mockAddress = {
  address1: "7742 Hwy 23",
  address2: "",
  country: "US",
  city: "Belle Chasse",
  firstName: "Salvos",
  lastName: "Seafood",
  postal: "70037",
  region: "LA",
  phone: "(504) 393-7303"
}

const mockMutation = (data) => new Promise((resolve, reject) => {
  setTimeout(() => {
    setState({
      cart: {
        fulfillmentGroup: {
          data: {
            shippingAddress: mockAddress
          }
        }
      }
    });
    resolve(mockAddress);
  }, 2000, { mockAddress });
});

const actions = [
  {
    label: "Shipping Information",
    component: ShippingAddressCheckoutAction,
    onSubmit: mockMutation,
    props: state.cart
  },
  { 
    label: "Second Shipping Information", 
    component: ShippingAddressCheckoutAction,
    onSubmit: mockMutation,
    props: {
      fulfillmentGroup: {
        data: {
          shippingAddress: mockAddress
        }
      }
    } 
  }
];


<CheckoutActions actions={actions} />

```
