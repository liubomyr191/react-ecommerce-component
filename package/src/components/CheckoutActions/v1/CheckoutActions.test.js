import React from "react";
import renderer from "react-test-renderer";
import mockComponents from "../../../tests/mockComponents";
import CheckoutActions from "./CheckoutActions";

class mockCheckoutAction extends React.Component {
  static renderComplete = () => <span />;
  render() {
    return <span />;
  }
}

const mockActions = [
  {
    label: "mock action one",
    status: "incomplete",
    component: mockCheckoutAction,
    onSubmit: () => true,
    props: {
      cartData: {
        data: null
      },
      cartMutation() {}
    }
  }
];

test("basic snapshot", () => {
  const component = renderer.create(<CheckoutActions components={mockComponents} actions={mockActions} />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
