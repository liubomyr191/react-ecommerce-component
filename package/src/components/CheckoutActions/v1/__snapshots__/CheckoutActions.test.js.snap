// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
.c0 {
  border-bottom: solid 1px #e6e6e6;
  padding: 1rem 0;
}

.c0:first-of-type {
  border-top: solid 1px #e6e6e6;
}

<div
  className="c0"
>
  CheckoutAction(undefined)
</div>
`;
